src/crypto/ocb_openssl.cc

```diff
@@ -1,1206 +1,178 @@
-            memcpy(next,ct,(size_t)i+16);
-            next = next+i+16;
-
-            ae_encrypt(&ctx,nonce,pt,first,pt,0,ct,NULL,AE_PENDING);
-            ae_encrypt(&ctx,NULL,pt+first,second,pt,0,ct+first,NULL,AE_PENDING);
-            ae_encrypt(&ctx,NULL,pt+first+second,third,pt,0,ct+first+second,NULL,AE_FINALIZE);
-            memcpy(next,ct,(size_t)i+16);
-            next = next+i+16;
-
-            ae_encrypt(&ctx,nonce,pt,0,pt,first,ct,NULL,AE_PENDING);
-            ae_encrypt(&ctx,NULL,pt,0,pt+first,second,ct,NULL,AE_PENDING);
-            ae_encrypt(&ctx,NULL,pt,0,pt+first+second,third,ct,NULL,AE_FINALIZE);
-            memcpy(next,ct,16);
-            next = next+16;
-        }
-
-    }
-    nonce[11] = 0;
-    ae_encrypt(&ctx,nonce,NULL,0,val_buf,next-val_buf,ct,tag,AE_FINALIZE);
-    pbuf(tag,16,0);
-    if (memcmp(valid,tag,16) == 0)
-    	printf("Vectors: PASS\n");
-    else
-    	printf("Vectors: FAIL\n");
-
-
-    /* Encrypt/Decrypt test */
-    for (i = 0; i < 128; i++) {
-        int first = ((i/3)/(BPI*16))*(BPI*16);
-        int second = first;
-        int third = i - (first + second);
-
-        nonce[11] = i%128;
-
-        if (1) {
-            len = ae_encrypt(&ctx,nonce,val_buf,i,val_buf,i,ct,tag,AE_FINALIZE);
-            len = ae_encrypt(&ctx,nonce,val_buf,i,val_buf,-1,ct,tag,AE_FINALIZE);
-            len = ae_decrypt(&ctx,nonce,ct,len,val_buf,-1,pt,tag,AE_FINALIZE);
-            if (len == -1) { printf("Authentication error: %d\n", i); return; }
-            if (len != i) { printf("Length error: %d\n", i); return; }
-            if (memcmp(val_buf,pt,i)) { printf("Decrypt error: %d\n", i); return; }
-        } else {
-            len = ae_encrypt(&ctx,nonce,val_buf,i,val_buf,i,ct,NULL,AE_FINALIZE);
-            ae_decrypt(&ctx,nonce,ct,first,val_buf,first,pt,NULL,AE_PENDING);
-            ae_decrypt(&ctx,NULL,ct+first,second,val_buf+first,second,pt+first,NULL,AE_PENDING);
-            len = ae_decrypt(&ctx,NULL,ct+first+second,len-(first+second),val_buf+first+second,third,pt+first+second,NULL,AE_FINALIZE);
-            if (len == -1) { printf("Authentication error: %d\n", i); return; }
-            if (memcmp(val_buf,pt,i)) { printf("Decrypt error: %d\n", i); return; }
-        }
-
-    }
-    printf("Decrypt: PASS\n");
+int ae_decrypt(ae_ctx *ctx, const void *nonce_ptr, const void *ct_ptr,
+	       int ct_len, const void *ad_ptr, int ad_len, void *pt_ptr,
+	       const void *tag, int final) {
+  assert(final == AE_FINALIZE);
+  int len = 0;
+  int plaintext_len = 0;
+  const unsigned char *nonce = (const unsigned char *)nonce_ptr;
+  const unsigned char *ct = (const unsigned char *)ct_ptr;
+  const unsigned char* ad = (const unsigned char *)ad_ptr;
+  unsigned char* pt = (unsigned char *)pt_ptr;
+  if (ct_len < ctx->tag_len) {
+    return AE_INVALID;
+  }
+  if (tag == NULL) {
+    ct_len -= ctx->tag_len;
+  }
```

Comment:
I remember this being somewhat non-obvious – perhaps it’s worth commenting about the default position of the tag here as well as in the encryption function?